rp2040-hal/src/multicore.rs

```diff
@@ -290,3 +331,68 @@ impl<'p> Core<'p> {
         }
     }
 }
+
+#[cfg(all(target_arch = "arm", feature = "thread_local"))]
+mod thread_local {
+    use core::arch::global_asm;
+    use core::ptr::addr_of;
+
+    extern "C" {
+        static TLS_CORE_0: u8;
+        static TLS_CORE_1: u8;
+    }
+    // Not really a const pointer, but we reform it into mut in the asm
+    static mut TLS_STATE: [*const u8; 2] = [
+        // Point to linker-allocated space in .bss
+        unsafe { addr_of!(TLS_CORE_0) },
+        unsafe { addr_of!(TLS_CORE_1) },
+    ];
+
+    // Define `__aeabi_read_tp` called by the compiler to get access to
+    // thread-local storage.
+    global_asm! {
+        ".pushsection .text.__aeabi_read_tp",
+        ".align 4",
+        ".p2align 4,,15",
+        ".global __aeabi_read_tp",
+        ".type __aeabi_read_tp,%function",
+
+        "__aeabi_read_tp:",
+        "    push {{r1, lr}}",
+        "    ldr r1, =0xd0000000",      // Load SIO CPUID addr
+        "    ldr r1, [r1]",             // Get current CPUID
+        "    lsls r1, r1, #2",          // Scale by 4
+        "    ldr r0, ={tls_state}",     // Load TLS_STATE base addr
+        "    ldr r0, [r0, r1]",         // Load CPU per-thread
+        "    pop {{r1, pc}}",
```

Comment:
The [Run-time ABI for the Arm® Architecture](https://github.com/ARM-software/abi-aa/blob/main/rtabi32/rtabi32.rst#thread-local-storage-new-in-v201) specifies that only `r0` may be clobbered by `__aeabi_read_tp`.

This won’t be a visible issue until someone starts enabling networking & tls operations.

Reply:
It's preserving r1 with push/pop. Am I missing something here?

Reply:
Oh right. 🤔 I guess it's still allowed to modify the stack pointer provided it restores it too.